rfcs/react-components/convergence/slot-children-render-function.md

```diff
@@ -0,0 +1,530 @@
+    </state.button>
+  </state.root>
+);
+```
+
+##### Option 3 stop mutating
+
+Let's just stop mutating, and since we require a layer of custom styling through classNames
+let's just properly use an argument on the render method exclusively for this:
+
+After:
+
+```tsx
+export const AccordionHeader: ForwardRefComponent<AccordionHeaderProps> = React.forwardRef((props, ref) => {
+  const state = useAccordionHeader_unstable(props, ref);
+  const styles = useAccordionHeaderStyles_unstable(state);
+  const contextValues = useAccordionHeaderContextValues_unstable(state);
+  return renderAccordionHeader_unstable(state, styles, contextValues);
+});
+
+export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState): AccordionHeaderStyles => {
+  const styles = useStyles();
+  return {
+    root: mergeClasses(
+      accordionHeaderClassNames.root,
+      styles.root,
+      state.inline && styles.rootInline,
+      state.disabled && styles.rootDisabled,
+      state.root.props.className,
+    ),
+  };
+};
+
+export const renderAccordionHeader_unstable = (state: AccordionHeaderState, styles: AccordionHeaderStyles) => (
+  <state.root className={styles.root}>
+    <state.button>
+      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
+      {state.icon && <state.icon />}
+      {state.root.props.children}
+      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
+    </state.button>
+  </state.root>
+);
+```
+
+#### Passing overrides from state to render
+
+Since our logic lies on the state hooks, the override methods will be implemented on those state hooks, although it should be
+declared on the render method explicitly to properly work as an override.
+
+For example, on the case `AccordionHeader` the `button` slot would override the `onClick` method. For that a property called `overrides`
+can be explicitly added to `AccordionHeaderState` stating such override:
+
+```tsx
+// AccordionHeader.types.ts
+type AccordionHeaderState = NextComponentState<AccordionHeaderSlots> & {
+  overrides: {
+    button: {
+      onClick(ev: React.MouseEvent<HTMLButtonElement>): void;
+    };
+  };
+};
+
+// useAccordionHeader.ts
+const state = {
+  // ...
+  overrides: {
+    button: {
+      onClick: useEventCallback(ev => {
+        buttonSlot.props.onClick?.(ev);
+        if (!ev.defaultPrevented) {
+          onAccordionHeaderClick(ev);
+        }
+      }),
+    },
+  },
+};
+
+// renderAccordionHeader.tsx
+const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
+  <state.root>
+    <state.button onClick={state.overrides.button.onClick}>
+      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
+      {state.icon && <state.icon />}
+      {state.root.props.children}
+      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
+    </state.button>
+  </state.root>
+);
+```
+
+#### Custom Pragma implementation
```

Comment:
how will this affect consumers ? 
- Will they have to switch to use our custom pragma in their codebase ( IMO that's a no go ).

For better context to reader can you please provide more info on how the outputs will look like and what's the consumers story gonna be with this custom pragma ?

Reply:
> how will this affect consumers ?

It won't (well, it'll allow the children render function support, that's the only difference for the consumers)

> Will they have to switch to use our custom pragma in their codebase ( IMO that's a no go ).

No. JSX pragma  is converted at build time and the consumers will consume the build output with the custom pragma already properly replaced.


> what's the consumers story gonna be with this custom pragma ?

There'll be no difference for consumers story!


> For better context to reader can you please provide more info on how the outputs will look like

👍🏻